use-case-1.groovy

```diff
@@ -0,0 +1,77 @@
+import hudson.model.*
+import jenkins.model.Jenkins
+import java.io.File;
+import groovy.xml.XmlUtil
+import static javax.xml.xpath.XPathConstants.*
+import java.io.InputStream;
+import java.io.FileInputStream
+import javax.xml.transform.stream.StreamSource
+import groovy.util.XmlParser
+
+def startingPoints=["Most_Granular_Level_Of_Jenkins_Folder"]
+def excludedPath = ["Project_3"]
+
+def allProject = []
+def excludedProjects = []
+
+Jenkins.instance.getAllItems(AbstractProject.class).each{item -> 
+  def startingPoint = startingPoints.each{ start -> 
+    if(item.fullName.toLowerCase().startsWith(start.toLowerCase())){
+      allProject.add(item.fullName.toLowerCase())
+    }
+  }
+}
+
+Jenkins.instance.getAllItems(AbstractProject.class).each{item -> 
+  def startingPoint = excludedPath.each{ start -> 
+    if(item.fullName.toLowerCase().contains(start.toLowerCase())){
+      excludedProjects.add(item.fullName.toLowerCase())
+    }
+  }
+}
+
+allProject.intersect(excludedProjects).each{allProject.remove(it);excludedProjects.remove(it)}
+
+def includedProject= allProject+excludedProjects
+
+def nexusIqBuilder = '''
+<org.sonatype.nexus.ci.iq.IqPolicyEvaluatorBuildStep plugin="nexus-jenkins-plugin@3.9.20200722-164144.e3a1be0">
+<com__sonatype__nexus__ci__iq__IqPolicyEvaluator____iqStage>Build</com__sonatype__nexus__ci__iq__IqPolicyEvaluator____iqStage>
+<com__sonatype__nexus__ci__iq__IqPolicyEvaluator____iqApplication class="org.sonatype.nexus.ci.iq.SelectedApplication">
+<applicationId></applicationId>
+</com__sonatype__nexus__ci__iq__IqPolicyEvaluator____iqApplication>
+<com__sonatype__nexus__ci__iq__IqPolicyEvaluator____failBuildOnNetworkError>false</com__sonatype__nexus__ci__iq__IqPolicyEvaluator____failBuildOnNetworkError>
+<com__sonatype__nexus__ci__iq__IqPolicyEvaluator____jobCredentialsId></com__sonatype__nexus__ci__iq__IqPolicyEvaluator____jobCredentialsId>
+<advancedProperties></advancedProperties>
+</org.sonatype.nexus.ci.iq.IqPolicyEvaluatorBuildStep>
+
+'''
+Jenkins.instance.getAllItems(AbstractProject.class).each {item ->
+  includedProject.each{include -> 
+    if(item.fullName.toLowerCase() == include.toLowerCase()){
+      def config = item.getConfigFile()
+      File file = config.getFile()
+      String fileContents = file.getText('UTF-8')
+      xml=new XmlParser().parseText(fileContents)
+      
+      def nodeToModify=xml.builders.findAll{n-> 
```

Comment:
I am not following how this works at all. `findAll(Closure c)` returns a collection of every element that `c` is true for.  Wouldn't this add the xml for every build step not named `"org.sonatype.nexus.ci.iq.IqPolicyEvaluatorBuildStep"`? Shouldn't we first check to see if the Project already has an IqPolicyEvaluatorBuildStep (by using `find`) and only adding it if we do not find one?

Reply:
Here is a simpler way of checking for the existence of the node and conditionally adding it
```groovy
NEXUS_IQ_BUILD_STEP_STR = '''
    <org.sonatype.nexus.ci.iq.IqPolicyEvaluatorBuildStep plugin="nexus-jenkins-plugin@3.9.20200722-164144.e3a1be0">
      <com__sonatype__nexus__ci__iq__IqPolicyEvaluator____iqStage>Build</com__sonatype__nexus__ci__iq__IqPolicyEvaluator____iqStage>
      <com__sonatype__nexus__ci__iq__IqPolicyEvaluator____iqApplication class="org.sonatype.nexus.ci.iq.SelectedApplication">
        <applicationId></applicationId>
      </com__sonatype__nexus__ci__iq__IqPolicyEvaluator____iqApplication>
      <com__sonatype__nexus__ci__iq__IqPolicyEvaluator____failBuildOnNetworkError>false</com__sonatype__nexus__ci__iq__IqPolicyEvaluator____failBuildOnNetworkError>
      <com__sonatype__nexus__ci__iq__IqPolicyEvaluator____jobCredentialsId></com__sonatype__nexus__ci__iq__IqPolicyEvaluator____jobCredentialsId>
      <advancedProperties></advancedProperties>
    </org.sonatype.nexus.ci.iq.IqPolicyEvaluatorBuildStep>
'''
NEXUS_IQ_BUILD_STEP = new XmlParser().parseText(NEXUS_IQ_BUILD_STEP_STR)

def project = new XmlParser().parse("Jenkins-Job-Before.xml")

if (! project.builders.'*'.find{ it.name() == 'org.sonatype.nexus.ci.iq.IqPolicyEvaluatorBuildStep'} ) {
    println "Adding Step"
    project.builders[0]?.append(NEXUS_IQ_BUILD_STEP)

    def sw = new StringWriter()
    new XmlNodePrinter(new PrintWriter(sw)).print(project)
    def modifiedXml = sw.toString()

    println modifiedXml
}
```

Reply:
This has been addressed in Inclusion-Exclusion-Array.groovy